Binders with default settings inherit from module's configuration

mbogoevici opened the pull request:
Resolves #229
Optimizes connection factory creation for binders that can use Boot's autoconfiguration

markfisher commented on spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/DefaultBinderFactory.java:
```diff
@@ -49,8 +51,8 @@ public DefaultBinderFactory(Map<String, BinderConfiguration> binderConfiguration
 	@Override
-	public void setEnvironment(Environment environment) {
-		this.environment = environment;
+	public void setApplicationContext(ApplicationContext applicationContext) throws BeansException {
```
minor thing, but this doesn't need the `throws` clause

mbogoevici replied:
I blame IntelliJ :)

markfisher commented on spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/DefaultBinderFactory.java:
```diff
@@ -110,12 +113,14 @@ else if (binderConfigurations.size() == 1) {
 							.sources(binderConfiguration.getBinderType().getConfigurationClasses())
 							.bannerMode(Mode.OFF)
 							.web(false);
+			if (binderProperties.isEmpty() && context != null) {
```
Might be worth a brief comment here to explain why we only set the parent context in this case.

mbogoevici replied:
Will add.

mbogoevici commented:
Rebased, added changes in https://github.com/spring-cloud/spring-cloud-stream/issues/232 and, additionally added support for suppressing environment inheritance (wherever necessary).

markfisher commented on spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/DefaultBinderFactory.java:
```diff
@@ -110,19 +113,30 @@ else if (binderConfigurations.size() == 1) {
+			}
+			else if (environment != null && binderConfiguration.isInheritEnvironment()) {
+				StandardEnvironment binderEnvironment = new StandardEnvironment();
+				environment.merge(environment);
```
This looks like a typo. I assume the intent is to merge into the `binderEnvironment`?

mbogoevici replied:
Yeah.

markfisher commented on spring-cloud-stream/src/test/java/org/springframework/cloud/stream/binder/BinderFactoryConfigurationTests.java:
```diff
@@ -90,6 +89,29 @@ public void loadBinderTypeRegistryWithOneBinder() throws Exception {
 	}
 
 	@Test
+	public void loadBinderTypeRegistryWithOneBinderAndSharedEnvironment() throws Exception {
```
this test and `loadBinderTypeRegistryWithOneBinderAndIsolatedEnvironment` make the same assertions - shouldn't we be testing something that varies with the boolean property?

mbogoevici replied:
Actually, they're not supposed to. Updating the tests now.

markfisher commented on spring-cloud-stream-binders/spring-cloud-stream-binder-redis/src/test/java/org/springframework/cloud/stream/binder/redis/integration/RedisBinderModuleTests.java:
```diff
@@ -0,0 +1,125 @@
+public class RedisBinderModuleTests {
+
+	@ClassRule
+	public static RedisTestSupport rabbitTestSupport = new RedisTestSupport();
```
wrong param name (copy paste) ;)

markfisher commented on spring-cloud-stream-binders/spring-cloud-stream-binder-redis/src/test/java/org/springframework/cloud/stream/binder/redis/integration/RedisBinderModuleTests.java:
```diff
@@ -0,0 +1,125 @@
+		RedisConnectionFactory binderConnectionFactory =
+				(RedisConnectionFactory) binderFieldAccessor.getPropertyValue("connectionFactory");
+		assertThat(binderConnectionFactory, instanceOf(RedisConnectionFactory.class));
+		RedisConnectionFactory connectionFactory = context.getBean(RedisConnectionFactory.class);
```
in the rabbit tests, the class passed to `getBean` is `ConnectionFactory.class` - why is this different?

markfisher commented on spring-cloud-stream-binders/spring-cloud-stream-binder-redis/src/test/java/org/springframework/cloud/stream/binder/redis/integration/RedisBinderModuleTests.java:
```diff
@@ -0,0 +1,125 @@
+		RedisConnectionFactory binderConnectionFactory =
+				(RedisConnectionFactory) binderFieldAccessor.getPropertyValue("connectionFactory");
+		assertThat(binderConnectionFactory, is(MOCK_CONNECTION_FACTORY));
+		RedisConnectionFactory connectionFactory = context.getBean(RedisConnectionFactory.class);
```
same question as in comment on line 76

markfisher commented on spring-cloud-stream-binders/spring-cloud-stream-binder-redis/src/test/java/org/springframework/cloud/stream/binder/redis/integration/RedisBinderModuleTests.java:
```diff
@@ -0,0 +1,125 @@
+		DirectFieldAccessor binderFieldAccessor = new DirectFieldAccessor(binder);
+		RedisConnectionFactory binderConnectionFactory =
+				(RedisConnectionFactory) binderFieldAccessor.getPropertyValue("connectionFactory");
+		RedisConnectionFactory connectionFactory = context.getBean(RedisConnectionFactory.class);
```
ditto

markfisher commented on spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/DefaultBinderFactory.java:
```diff
@@ -51,8 +51,8 @@ public DefaultBinderFactory(Map<String, BinderConfiguration> binderConfiguration
 	@Override
-	public void setEnvironment(Environment environment) {
-		this.environment = environment;
+	public void setApplicationContext(ApplicationContext applicationContext) {
```
should this have an `Assert.instanceOf(..)` prior to the cast?

markfisher commented:
will address my 2 last comments while merging

markfisher commented:
merged along with https://github.com/spring-cloud/spring-cloud-stream/commit/df1817cc518eb979d49ab1030fada5ec4091cd8a
